components/search/src/selector.rs

```diff
@@ -1,27 +1,60 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-use crate::{error::Error, SearchApiResult};
+
+use crate::{error::Error, RefinedConfig, SearchApiResult, SearchUserEnvironment};
 use error_support::handle_error;
 
+/// SearchEngineSelector parses the JSON configuration for
+/// search engines and returns the applicable engines depending
+/// on their region + locale.
+#[derive(uniffi::Object)]
 pub struct SearchEngineSelector();
 
+#[uniffi::export]
 impl SearchEngineSelector {
+    /// Sets the search configuration from the given string. This allows for
+    /// reprocessing of the configuration if it has not changed since the
+    /// previous update, e.g. to optimise test running.
+    #[handle_error(Error)]
+    pub fn set_search_config(&self, _configuration: String) -> SearchApiResult<()> {
+        Err(Error::NotImplemented)
+    }
+
+    /// Clears the search configuration from memory if it is known that it is
+    /// not required for a time.
+    pub fn clear_search_config(&self) {}
+
+    /// Filters the search configuration with the user's given environment,
+    /// and returns the set of engines and parameters that should be presented
+    /// to the user.
     #[handle_error(Error)]
-    pub fn filter_engine_configuration(&self) -> SearchApiResult<()> {
+    pub fn filter_engine_configuration(
```

Comment:
Did we finalize if this would be an ordered set of search engines or not?

Reply:
At the moment it does, but I'm in two minds about this. On desktop, we don't use the order of the engines from the selector because when we order the engines, we take into account the user installed engines. So the default order on desktop is roughly:

* Default Engine
* Engines defined in the `orders` section of the configuration
* All other engines (application and user-installed) in alphabetical order

Obviously the engine selector can't manage that last point as it doesn't know about user installed engines.

I think mobile currently does:

* Default Engine
* All other application engines in alphabetical order
* All user installed engines in alphabetical order

Desktop also allows reordering of the engines by the user.

Overall, I'm not sure if the selector should actually handle the order, since that feels more of a UI function than back-end functionality, however, I'm open to thoughts.

Reply:
Hmm, this is indeed a bit of a weird case. IMO, generally speaking, I think this kind of comes down to "does the search engine order 'matter' to the Search team?" because, if we wind up having to make 3x the changes every time we decide to change the search engine order, that feels like the type of problem this consolidation project is meant to fix.

But if it's more-or-less "fine" to let each of the clients make different business decisions about search engine ordering, then it'd be fine to leave this unsorted.

IMO, perhaps a middle ground would be a "loose" API contract about Search engine order, which I imagine covers the majority of user scenarios - e.g., perhaps this *should* promise to return them in order, but with a clarifying bit of API documentation saying "list is sorted in default search engine order, but clients might override that (e.g. to accommodate user-entered search engines, or to allow users to manually change the order)."

Reply:
I've just updated the PR with the new comments about the search order.

I also added an `orderHint` field to `SearchEngineDefinition` as I realised that it was missing, and desktop would need it.